### ISSUE TITLE
Speedup symmetric eigendecomposition

### ISSUE THREAD

**Issue by t4c1** [stan-dev/math#1156]
## Description

Symmetric eigendecomposition from library Eigen is implemented in a suboptimal way. It consists of two steps: tridiagonalization and tridiagonal eigensolver.

In Eigen, basic Householder tridiagonalization is implemented. This can be improved on using blocked Householder algorithm, which uses cache more efficiently and is therefore faster.

For symmetric eigensolver Eigen uses QR iteration. While very accurate, it is quite slow. For square symmetric matrix of size n it does O(n^3) operations. It can be replaced by, for example, the
MRRR algorithm that is significantly faster - O(n^2) and usually accurate enough. However, for cases where multiple eigenvalues are very close to each other - with relative difference (l1-l2)/l1 between
them close to machine precision - MRRR might produce more significant errors or in extreme cases even fail to converge.

Both steps can be further improved using GPUs.

## Example

Function is added that computes symmetric eigendecomposition.

## Expected Output

Symmetric eigendecomposition is faster than one implemented in Eigen.

#### Current Version:
v2.18.1

**Comment by t4c1** [comment]
I made something similar to your test: https://gist.github.com/t4c1/b8112f4bb9b3d6261f29cb578fa20090

**Comment by wds15** [comment]
Could u please plot that table? I am bad in mentally plotting it...

...this looks quite nice...

**Comment by t4c1** [comment]
Here are the graphs for times and errors:
![CPU speedup](https://user-images.githubusercontent.com/7499452/54439917-0b9ca000-473a-11e9-8114-cb9f0c76cefe.png)
![CPU_errors](https://user-images.githubusercontent.com/7499452/54439925-0dfefa00-473a-11e9-93da-7e75bd5bb200.png)

**Comment by SteveBronder** [comment]
Nice! `bstatcomp:cpu_eigendecomposition` is the branch I should be on to test the above?

**Comment by rok-cesnovar** [comment]
Yep: https://github.com/bstatcomp/math/tree/cpu_eigendecomposition

There is also a variant for OpenCL where I think Tadej got around a 5-6x speedup compared to this branch on a GTX1070.
